File: test/corpus/propagation/cast_struct_literal_field.exs
defmodule RustQ.Corpus.Propagation.CastStructLiteralField do
  @moduledoc "Struct literal field checking propagates through cast/2 operands."

  use RustQ.Meta

  alias RustQ.Type, as: R

  defmodule Point do
    defstruct [:x]
  end

  @type point :: %Point{x: R.f32()}

  @spec decode_i64(term()) :: R.nif_result(R.i64())
  defrust decode_i64(term) do
    decode_as!(term, R.i64())
  end

  @spec consume(point()) :: R.nif_result(R.unit())
  defrust consume(_point) do
    :ok
  end

  @spec run(term()) :: R.nif_result(R.unit())
  defrust run(term) do
    point = struct_literal(Point, x: cast(decode_i64(term), R.f32()))
    consume(point)
    :ok
  end
end

File: test/corpus/propagation/cast_struct_literal_field.rs
#[derive(Clone, Debug)]
pub struct Point {
    pub x: f32,
}

pub fn decode_point<'a>(term: Term<'a>) -> NifResult<Point> {
    Ok(Point {
        x: term.map_get(atoms::x())?.decode()?,
    })
}

fn decode_i64<'a>(term: Term<'a>) -> NifResult<i64> {
    term.decode::<i64>()?
}

fn consume(_point: Point) -> NifResult<()> {
    Ok(())
}

fn run<'a>(term: Term<'a>) -> NifResult<()> {
    let point = Point {
        x: decode_i64(term)? as f32,
    };
    consume(point)?;
    Ok(())
}
